src/framework/global/types/number.h

```diff
@@ -0,0 +1,171 @@
+#ifndef MU_GLOBAL_NUMBER_T_H
+#define MU_GLOBAL_NUMBER_T_H
+
+#include <limits>
+#include <cassert>
+#include <cmath>
+
+#include "global/logstream.h"
+
+namespace mu {
+// default
+constexpr double COMPARE_REAL_EPSILON(1000000.0);
+constexpr double COMPARE_REAL_NULL(0.000001);
+
+// use
+inline double _compare_real_epsilon(COMPARE_REAL_EPSILON);
+inline double _compare_real_null(COMPARE_REAL_NULL);
+
+template<typename T>
+inline bool is_zero(T v)
+{
+    if constexpr (std::numeric_limits<T>::is_integer) {
+        return v == 0;
+    } else {
+        return std::abs(v) <= _compare_real_null;
+    }
+}
+
+template<typename T>
+inline bool is_equal(T v1, T v2)
+{
+    if constexpr (std::numeric_limits<T>::is_integer) {
+        return v1 == v2;
+    } else {
+        return std::abs(v1 - v2) * _compare_real_epsilon <= std::min(std::abs(v1), std::abs(v2));
+    }
+}
+
+template<typename T>
+class number_t
+{
+public:
+
+    inline T check_valid(T v, T def = T())
+    {
+        bool valid = !std::isnan(v) && !std::isinf(v);
```

Comment:
Just a friendly reminder. This won't work with -Ofast optimization

Reply:
Mainly it needs to work in debug :)